speechbrain/lobes/models/discrete/sq_codec.py

```diff
@@ -0,0 +1,1394 @@
+            weight_norm(
+                Conv1d(
+                    init_channel,
+                    num_bands,
+                    kernel_size=default_kernel_size,
+                    causal=causal,
+                )
+            )
+        )
+
+        self.encoder = nn.ModuleList(self.encoder)
+        self.decoder = nn.ModuleList(self.decoder)
+
+    def forward(self, x):
+        """
+        Performs a forward pass through the encoder and decoder.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor of shape (batch_size, channels, length).
+
+        Returns
+        -------
+        torch.Tensor
+            Reconstructed output tensor.
+        """
+        for i, layer in enumerate(self.encoder):
+            if i != len(self.encoder) - 1:
+                x = layer(x)
+            else:
+                x = F.tanh(layer(x))
+        x = self.vq(x)  # Quantization step
+        for i, layer in enumerate(self.decoder):
+            x = layer(x)
+        return x
+
+    def inference(self, x):
+        """
+        Encodes input tensor `x` and decodes the quantized embeddings.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor of shape (batch_size, channels, length).
+
+        Returns
+        -------
+        tuple
+            A tuple (emb, emb_quant, x), where `emb` is the latent embedding,
+            `emb_quant` is the quantized embedding, and `x` is the decoded output.
+        """
+        for i, layer in enumerate(self.encoder):
+            if i != len(self.encoder) - 1:
+                x = layer(x)
+            else:
+                x = F.tanh(layer(x))
+        emb = x
+        emb_quant = self.vq(emb)
+        x = emb_quant
+        for i, layer in enumerate(self.decoder):
+            x = layer(x)
+        return emb, emb_quant, x
+
+    def encode(self, x):
+        """
+        Encodes the input tensor `x` into a quantized embedding.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor of shape (batch_size, channels, length).
+
+        Returns
+        -------
+        torch.Tensor
+            Quantized embedding.
+        """
+        for i, layer in enumerate(self.encoder):
+            if i != len(self.encoder) - 1:
+                x = layer(x)
+            else:
+                x = F.tanh(layer(x))
+        emb = x
+        emb_quant = self.vq(emb)
+        return emb_quant
+
+    def decode(self, emb_quant):
+        """
+        Decodes the quantized embeddings back into a tensor.
+
+        Parameters
+        ----------
+        emb_quant : torch.Tensor
+            Quantized embedding tensor.
+
+        Returns
+        -------
+        torch.Tensor
+            Reconstructed output tensor.
+        """
+        x = emb_quant
+        for i, layer in enumerate(self.decoder):
+            x = layer(x)
+        return x
+
+
+class CustomRoundingFunction(Function):
```

Comment:
Since it only used for this module so far, lets keep it here. I am working on another recepie that include many of these quantization techniques.. i will take care of this in that PR.